File: Data Structures/6. Queue/Regular Queue/arrayQueue.cpp
// a simple implementation with vector
#include <iostream>
#include <vector>
using namespace std;

class ArrayQueue {
  public:
  vector<int> queue;
  int front() {return queue.front();} // peek the front of the queue
  int isEmpty() {return queue.empty();} // check if the queue is empty
  int size() {return queue.size();} // return the size of the queue
  void enqueue(int value) {queue.push_back(value);} // push value to back
  void dequeue() {if (!isEmpty()) queue.erase(queue.begin());} // pop oldest value
  void output() {
    // empty, no need for output
    if (queue.empty()) {
      cout << "queue is empty!" << endl;
      return;
    }

    // print out the values
    cout << "contents: ";
    vector<int> copy = queue;
    while (!copy.empty()) {
      cout << copy.front() << " ";
      copy.erase(copy.begin());
    }
    cout << endl;
  }
};

// print function
void print(string text) {cout << text;}

// test the queue
int main() {
  print("Array implementation of queue\n\n");

  // create a queue
  ArrayQueue queue;
  print("-queue created!\n");
  queue.output(); // output if content exists
  print("\n");

  // push value into the queue
  queue.enqueue(1);
  queue.enqueue(2);
  queue.enqueue(3);
  print("-values are enqueued into the queue!\n");
  queue.output(); // output if content exists
  print("\n");

  // pop value from the top
  queue.dequeue();
  print("-dequeue the queue!\n");
  queue.output(); // output if content exists
  print("\n");

  // size of the staclk
  cout << "-The size is " << queue.size() << "\n";
  queue.output(); // output if content exists
  print("\n");
}

File: Data Structures/6. Queue/Regular Queue/arrayQueue.py
# simple implementation with list
class ArrayQueue:
  # constructor
  def __init__(self):
    self.queue = []

  # check if the queue is empty
  def is_empty(self):
    return len(self.queue) == 0

  # peek the front of the queue
  def front(self):
    # if empty
    if self.is_empty():
      print("The queue is empty!")
    return self.queue[0] # not empty, peek it

  # enqueue value into the queue
  def enqueue(self, value):
    self.queue.append(value)
  
  # dequeue the top of the queue
  def dequeue(self):
    # if empty
    if self.is_empty():
      print("The queue is empty!")
      return
    self.queue.remove(self.queue[0]) # dequeue the front of the queue

  # size of the queue
  def size(self):
    return len(self.queue)
  
  # output the queue
  def output(self):
    if self.is_empty():
      print("queue is empty!")
    print("contents: "),
    print(self.queue[:])
    print

# ============================
# testing the queue class
# ============================

print("\nList implementation of queue!\n")

# create a new ArrayStack object
queue = ArrayQueue()
print("- New ArrayQueue object created")
queue.output()

# enqueue value into the queue
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
print("- New values are pushed into the queue")
queue.output()

# peek value
print("- Peek the front of the queue: " + str(queue.front()))
print

# pop the top of the queue
queue.dequeue()
print("- Pop the top of the queue")
queue.output()

# get size of the queue
print("- Size of the queue: " + str(queue.size()))
print